llvm/lib/IR/BasicBlock.cpp

```diff
@@ -1030,6 +1037,7 @@ void BasicBlock::spliceDebugInfoImpl(BasicBlock::iterator Dest, BasicBlock *Src,
     DPMarker *TrailingDPValues = getTrailingDPValues();
     if (TrailingDPValues) {
       FirstMarker->absorbDebugValues(*TrailingDPValues, true);
+      TrailingDPValues->eraseFromParent();
```

Comment:
Should `absorbDebugValues` call `eraseFromParent`? Rather than calling it all the `absorbDebugValues` call sites. AFAICT absorb always absorbs all the DPValues, so it would probably be a rare case that we want to keep the absorbed-from marker around?

Reply:
Eventually yes -- however the RemoveDIs/DDD model in-tree right now has one DPMarker allocated per instruction, and we haven't landed the plumbing required to support that (we'll get null pointer exceptions everywhere otherwise).

Over in 2f93a721eaf2aa348, which is a refactoring of the internal patch we've got, there's an "adopt" call that works very similarly to what you describe here. It might be able to automate erasing (erasal?) but still requires juggling with the trailingDPValues :(.